sources/bootstrap-commands/src/main.rs

```diff
@@ -0,0 +1,367 @@
+mode = "always"
+
+This would set `/etc/motd` to "helloworld".
+
+# Additional Information:
+Certain valid apiclient commands that work in a session may fail in bootstrap-commands due to
+relevant services not running at the time of the launch of the systemd service.
+
+## Example:
+[settings.bootstrap-commands.001-test-bootstrap-commands]
+commands = [[ "apiclient", "exec", "admin", "ls"]]
+essential = true
+mode = "always"
+
+This command fails because bootstrap-commands.service which calls this binary is launched prior
+to preconfigured.target while host-containers@.service which is a requirement for running "exec"
+commands are launched after preconfigured.target.
+*/
+
+use log::info;
+use serde::Deserialize;
+use simplelog::{Config as LogConfig, LevelFilter, SimpleLogger};
+use snafu::{ensure, OptionExt, ResultExt};
+use std::collections::BTreeMap;
+use std::env;
+use std::ffi::OsStr;
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::process::{self, Command};
+use std::str::FromStr;
+
+use bottlerocket_modeled_types::{ApiclientCommand, BootstrapMode, Identifier};
+
+const DEFAULT_CONFIG_PATH: &str = "/etc/bootstrap-commands/bootstrap-commands.toml";
+
+#[derive(Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct BootstrapCommandConfig {
+    #[serde(default)]
+    bootstrap_commands: BTreeMap<Identifier, BootstrapCommand>,
+}
+
+impl BootstrapCommandConfig {
+    // Gets an iterator for bootstrap_commands, sorted in lexicographical order of their names.
+    fn iter(self) -> impl Iterator<Item = (Identifier, BootstrapCommand)> {
+        self.bootstrap_commands.into_iter()
+    }
+}
+
+#[derive(Debug, PartialEq, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct BootstrapCommand {
+    #[serde(default)]
+    commands: Vec<ApiclientCommand>,
+    #[serde(default)]
+    mode: BootstrapMode,
+    #[serde(default)]
+    essential: bool,
+}
+
+/// Stores user-supplied global arguments
+struct Args {
+    log_level: LevelFilter,
+    config_path: PathBuf,
+}
+
+impl Default for Args {
+    fn default() -> Self {
+        Self {
+            log_level: LevelFilter::Info,
+            config_path: PathBuf::from_str(DEFAULT_CONFIG_PATH).unwrap(),
+        }
+    }
+}
+
+/// Wrapper around process::Command that adds error checking.
+fn run_command<I, S>(bin_path: &str, args: I) -> Result<()>
+where
+    I: IntoIterator<Item = S>,
+    S: AsRef<OsStr>,
+{
+    let mut command = Command::new(bin_path);
+
+    command
+        .args(args)
+        .status()
+        .context(error::ExecutionFailureSnafu { command })?;
+
+    Ok(())
+}
+
+fn handle_bootstrap_command<S>(name: S, bootstrap_command: BootstrapCommand) -> Result<()>
+where
+    S: AsRef<str>,
+{
+    let name = name.as_ref();
+    let mode = bootstrap_command.mode.as_ref();
+    let commands = &bootstrap_command.commands;
+
+    if mode == "off" {
+        // If mode is 'off', just log this information.
+        info!("Bootstrap command mode for '{}' is 'off'", name);
```

Comment:
nit: can do an early return to outdent the remaining code:
```suggestion
    if mode == "off" {
        // If mode is 'off', just log this information.
        info!("Bootstrap command mode for '{}' is 'off'", name);
        return  Ok(());
    }
```